ast/check.go

```diff
@@ -146,7 +146,97 @@ func (tc *typeChecker) checkLanguageBuiltins(env *TypeEnv, builtins map[string]*
 }
 
+func getPrefixToOverride(name string, env *TypeEnv, rule *Rule) (string, types.Type, *Error) {
+	ref, err := ParseRef(name)
+	if err != nil {
+		return "", nil, NewError(TypeErr, rule.Location, err.Error())
+	}
+	t := env.tree.Get(ref)
+	if t != nil {
+		return name, t, nil
+	}
+	names := strings.Split(name, ".")
+	if len(names) > 1 {
+		prefix, t, err := getPrefixToOverride(strings.TrimSuffix(name, "."+names[len(names)-1]), env, rule)
+		if err != nil {
+			return "", nil, err
+		}
+		if t != nil {
+			return prefix, t, nil
+		}
+	}
+	return "", nil, nil
+}
+
+func override(names []string, t types.Type, o types.Type) types.Type {
+	obj := t.(*types.Object)
+	staticProps := obj.StaticProperties()
+	newStaticProps := []*types.StaticProperty{}
+
+	for _, prop := range staticProps {
+		if len(names) == 1 && names[0] == prop.Key.(string) {
+			prop.Value = o
+		} else if len(names) > 0 && names[0] == prop.Key.(string) {
+			prop.Value = override(names[1:], prop.Value, o)
+		}
+		newStaticProps = append(newStaticProps, types.NewStaticProperty(prop.Key, prop.Value))
+	}
+	return types.NewObject(newStaticProps, nil)
+}
+
+// Annotations must immediately precede the rule definition and are of the form: #@rulesSchema=<expr>:<schema-key>
+func (tc *typeChecker) processAnnotation(annot *SchemaAnnotation, env *TypeEnv, rule *Rule) (Ref, types.Type, *Error) {
+	if env.schemaSet == nil || env.schemaSet.ByPath == nil {
+		return nil, nil, NewError(TypeErr, rule.Location, "Schemas need to be supplied for the annotation: %s", annot.Schema)
+	}
+	schema, ok := env.schemaSet.ByPath[annot.Schema]
+	if !ok {
+		return nil, nil, NewError(TypeErr, rule.Location, "Schema does not exist for given path in annotation: %s", annot.Schema)
+	}
+	newType, err := setTypesWithSchema(schema)
+	if err != nil {
+		return nil, nil, NewError(TypeErr, rule.Location, err.Error())
+	}
+	ref, err := ParseRef(annot.Name)
+	if err != nil {
+		return nil, nil, NewError(TypeErr, rule.Location, err.Error())
+	}
+
+	return ref, newType, nil
+}
+
 func (tc *typeChecker) checkRule(env *TypeEnv, rule *Rule) {
+	if rule.Annotations != nil {
+		errors := []*Error{}
+		for _, annot := range rule.Annotations {
+			ref, refType, err := tc.processAnnotation(annot, env, rule)
+			if err != nil {
+				errors = append(errors, err)
+				continue
+			}
+			prefixName, t, err := getPrefixToOverride(annot.Name, env, rule)
+			if err != nil {
+				errors = append(errors, err)
+				continue
+			}
+			if t == nil || prefixName == annot.Name {
+				env.tree.Put(ref, refType)
+				defer env.tree.DeleteKey(ref)
+			} else {
+				refPrefix, err := ParseRef(prefixName)
+				if err != nil {
+					errors = append(errors, NewError(TypeErr, rule.Location, err.Error()))
+					continue
+				}
+				name := strings.TrimPrefix(annot.Name, prefixName+".")
+				names := strings.Split(name, ".")
+				newType := override(names, t, refType)
+				env.tree.Put(refPrefix, newType)
+				defer env.tree.Put(refPrefix, t)
```

Comment:
I'm confused that we're calling `Put/defer DeleteKey` above and `Put/defer Put` here. 🤔 

Do we have test cases covering this? 🔍

Reply:
Added tests in commit https://github.com/open-policy-agent/opa/pull/3123/commits/3eb2e9f68305c617c8020dc8a984d36cca2be2e4

We need to do defers and remove the changes made the the type environment after the CheckRule is done. This is because the schema annotation is only is scope within the rule. So we modify the type env according to the schema annotation before checking the rule, and remove the changes after checking the rule.

Reply:
@srenatus we removed the defers in favor of env.wrap()